File: C++/longestValidParentheses.cpp
/**
 * @File    : longestValidParentheses.cpp
 * @Brief   : 最长有效括号
 * @Link    : https://leetcode-cn.com/problems/longest-valid-parentheses/
 * @Author  : Wei Li
 * @Date    : 2021-08-17
*/

#include <iostream>
#include <string>
#include <algorithm>

class Solution
{
public:
    int longestValidParentheses(std::string s)
    {
        int left = 0, right = 0, maxlength = 0;
        for (int i = 0; i < s.length(); i++)
        {
            if (s[i] == '(')
            {
                left++;
            }
            else
            {
                right++;
            }
            if (left == right)
            {
                maxlength = std::max(maxlength, 2 * right);
            }
            else if (right > left)
            {
                left = right = 0;
            }
        }
        left = right = 0;
        for (int i = (int)s.length() - 1; i >= 0; i--)
        {
            if (s[i] == '(')
            {
                left++;
            }
            else
            {
                right++;
            }
            if (left == right)
            {
                maxlength = std::max(maxlength, 2 * left);
            }
            else if (left > right)
            {
                left = right = 0;
            }
        }
        return maxlength;
    }
};

// ------------------------------
int main(int argc, char **argv)
{
    // std::string str = "(()";
    std::string str = ")()())";

    Solution solution;
    int longest_valid_parenttheses = solution.longestValidParentheses(str);
    std::cout << "The solution of this problem is " << longest_valid_parenttheses << std::endl;

    return 0;
}

File: Python/longestValidParentheses.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: longestValidParentheses.py
@Function: 最长有效括号
@Link: https://leetcode-cn.com/problems/longest-valid-parentheses/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-17
"""


from typing import List


class Solution:
    def longestValidParentheses(self, s: str) -> int:
        if not s:
            return 0
        res = 0
        stack = [-1]
        for i in range(len(s)):
            if s[i] == "(":
                stack.append(i)
            else:
                stack.pop()
                if not stack:
                    stack.append(i)
                else:
                    res = max(res, i - stack[-1])
        return res


# -------------------------
if __name__ == "__main__":
    str = ")()())"

    solution = Solution()
    longest_valid_parenttheses = solution.longestValidParentheses(str)
    print(f"The solution of this problem is {longest_valid_parenttheses}")
